### ISSUE TITLE
Basis member values are different in c#

### ISSUE THREAD

**Issue by khairul169** [godotengine/godot#11901]
**Operating system or device, Godot version, GPU Model and driver (if graphics related):**
Windows 10, Godot 3.0 master 195b12212d96494c3c7260c44442fcdfe27060e5

**Issue description:**

Member values of basis are different than using gdscript.. here is the output of sample project:

```
CS: (0) (1, 0, 0)
CS: (30) (0.866025, 0, 0.5)
CS: (60) (0.5, 0, 0.866025)
CS: (90) (-0, 0, 1)
CS: (120) (-0.5, 0, 0.866025)
CS: (150) (-0.866025, 0, 0.5)

GD: (0) (1, 0, 0)
GD: (30) (0.866025, 0, -0.5)
GD: (60) (0.5, 0, -0.866025)
GD: (90) (-0, 0, -1)
GD: (120) (-0.5, 0, -0.866025)
GD: (150) (-0.866025, 0, -0.5)
```

gdscript:
```
for i in range(6):
	var deg = 30*i;
	var basis = Basis(Vector3(0, 1, 0), deg2rad(deg));
	print("GD: (", deg, ") ", basis.x);
```

c#:
```
for (int i = 0; i < 6; i++)
{
	float deg = 30.0f * i;
	Basis basis = new Basis(new Vector3(0, 1, 0), Mathf.deg2rad(deg));

	GD.print("CS: (", deg, ") ", basis.x);
}
```

**Steps to reproduce:**

**Link to minimal example project:**

[TestBasis.zip](https://github.com/godotengine/godot/files/1364291/TestBasis.zip)

**Comment by khairul169** [comment]
Maybe #11899 fixes this issue? 😄

**Comment by tagcup** [comment]
No, this is appears to be a different issue.

Can you try basis.get_axis(0) in C# and GDScript?

**Comment by tagcup** [comment]
In GDScript, basis.x apparently returns Basis::get_axis(0) (which is a C++ function) which returns the first column of the basis matrix.

In C#, Basis is stored as 3 vectors, each representing a *row*, and .x gives you the first row.

This is an API inconsistency which should be addressed, so could you re-open the issue?

**Comment by khairul169** [comment]
Okay, finally returned what it should by using basis.get_axis(0) in c# and basis[0] in gdscript.. It looks like the .x and [0] issue in the c# itself..
```
C#:
[30] basis: ((0.866025, 0, 0.5), (0, 1, 0), (-0.5, 0, 0.866025)) | basis[0]: (0.866025, 0, -0.5)
[60] basis: ((0.5, 0, 0.866025), (0, 1, 0), (-0.866025, 0, 0.5)) | basis[0]: (0.5, 0, -0.866025)
[90] basis: ((-0, 0, 1), (0, 1, 0), (-1, 0, -0)) | basis[0]: (-0, 0, -1)

GDScript:
[30] basis: ((0.866025, 0, 0.5), (0, 1, 0), (-0.5, 0, 0.866025)) | basis[0]: (0.866025, 0, -0.5)
[60] basis: ((0.5, 0, 0.866025), (0, 1, 0), (-0.866025, 0, 0.5)) | basis[0]: (0.5, 0, -0.866025)
[90] basis: ((-0, 0, 1), (0, 1, 0), (-1, 0, -0)) | basis[0]: (-0, 0, -1)
```

Edited scripts:
[TestBasis.zip](https://github.com/godotengine/godot/files/1366622/TestBasis.zip)

**Comment by tagcup** [comment]
The member variables (x, y, z) in C# are badly named, unfortunately. They don't represent object basis vectors.

**Comment by tagcup** [comment]
Probably language limitation, C# apparently doesn't allow arrays with fixed lengths, like float elements[3][3] isn't valid C#.

**Comment by khairul169** [comment]
So, `basis[0]`, in gdscript returned vector3 of the axis, and in c# one, returned the row.. Is this a known issue? If yes, probably linearizing with the same method as gdscript is needed, or at least documenting it somewhere to prevent misunderstanding like this..

**Comment by tagcup** [comment]
The best solution is probably to rename C# member variables. There is no .x .y .z in the original C++ reference code anyway, had no idea they existed in GDScript until you mentioned.

**Comment by tagcup** [comment]
Renaming those vector to m0, m1, m2 for example in C# would make things read correctly, like m1[2] would give the M_12 element of the matrix.

**Comment by Zylann** [comment]
C# supports [3,3], I think (with a comma).

**Comment by tagcup** [comment]
I don't mean with new but as a part of the the type, static allocation. There seems to be fixed keyword for that but is for unsafe code

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/unsafe-code-pointers/fixed-size-buffers

**Comment by neikeq** [comment]
This should have been fixed by #17360. Not sure about #14553. I will check as soon as I can.

**Comment by Chaosus** [comment]
(checked) Yeah, its fixed.
